README.md
# libhashring

This library provides a consistent hashing implementation that supports SHA-1 and MD5. It is high performance and suitable for rings with a large number of items (replicas or nodes). The following bindings are available:

 * Erlang
 * Java

## License (See LICENSE file for full license)

Copyright 2011 LiveProfile

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

## Overview

**hash_ring** exposes a simple API to create/remove rings, add/remove nodes, and locate the node for a key on the ring.

A ring is created by using the *hash_ring_create()* function.

*numReplicas*, which determines how many times a node is placed on the ring. This parameter gives you flexibility on how consistent the ring is. Larger values will even out the node distribution on the ring. 128 is a sensible default for most users.

*hash_fn* is the hash function to use. Currently *HASH_FUNCTION_MD5* and *HASH_FUNCTION_SHA1* are supported. In tests MD5 is on average about 25% faster.

    hash_ring_t *ring = hash_ring_create(128, HASH_FUNCTION_SHA1);

Next you can add some nodes to the ring. For example, if you have a cluster of Redis servers, you might do the following:

    char *redis01 = "redis01", *redis02 = "redis02", *redis03 = "redis03";

    hash_ring_add_node(ring, (uint8_t*)redis01, strlen(redis01));
    hash_ring_add_node(ring, (uint8_t*)redis02, strlen(redis02));
    hash_ring_add_node(ring, (uint8_t*)redis03, strlen(redis01));
    
The ring will now have **384** items, **128** per node (with 3 nodes total).

## Node hashing

It is helpful to know how a node is hashed onto the ring, especially if you want to write a compatible library for another language or platform.

**hash_ring** hashes the node's name followed by the ASCII number of the replica. For example, with *2* replicas, the following pseudo code is used to hash the node:

    redis01_0_int = SHA1("redis01" + "0") & 0x000000000000000000000000ffffffffffffffff
    redis01_1_int = SHA1("redis01" + "1") & 0x000000000000000000000000ffffffffffffffff
  
The least significant **64-bits** of the digest are used to retrieve the integer.

**Note**: You can use *hash\_ring\_set\_mode* to use HASH\_RING\_MODE\_LIBMEMCACHED\_COMPAT which will add nodes as "node-index" and will hash to a 32-bit integer instead. This is what libmemcached uses.

## Compiling 

Compile the library and install with:

    sudo make install
    
This will install the library to **/usr/local/lib/libhashring.so**. The header file *hash_ring.h* is copied  to **/usr/local/include/hash_ring.h**.

## Running the tests

    make test
    
## Bindings

To build all the bindings just run:

    make bindings

## Example

The following code is from the tests, which hashes some known keys and ensures that they are located on the ring properly.

    hash_ring_t *ring = hash_ring_create(8, HASH_FUNCTION_SHA1);
    char *slotA = "slotA";
    char *slotB = "slotB";

    char *keyA = "keyA";
    char *keyB = "keyBBBB";
    char *keyC = "keyB_";

    hash_ring_node_t *node;

    assert(hash_ring_add_node(ring, (uint8_t*)slotA, strlen(slotA)) == HASH_RING_OK);
    assert(hash_ring_add_node(ring, (uint8_t*)slotB, strlen(slotB)) == HASH_RING_OK);

    node = hash_ring_find_node(ring, (uint8_t*)keyA, strlen(keyA));
    assert(node != NULL && node->nameLen == strlen(slotA) && memcmp(node->name, slotA, strlen(slotA)) == 0);

    node = hash_ring_find_node(ring, (uint8_t*)keyB, strlen(keyB));
    assert(node != NULL && node->nameLen == strlen(slotA) && memcmp(node->name, slotA, strlen(slotA)) == 0);

    node = hash_ring_find_node(ring, (uint8_t*)keyC, strlen(keyC));
    assert(node != NULL && node->nameLen == strlen(slotB) && memcmp(node->name, slotB, strlen(slotB)) == 0);

    hash_ring_free(ring);

## Erlang

If you use *rebar* you can put the following in your *rebar.config* to use libhashring:
	
	{deps, [
	 {hash_ring, ".*", {git, "https://github.com/chrismoos/hash-ring.git", "master"}}
	]}.

### Example

    hash_ring:start_link(),
    Ring = "myring",
    ?assert(hash_ring:create_ring(Ring, 8) == ok),
    ?assert(hash_ring:add_node(Ring, <<"slotA">>) == ok),
    ?assert(hash_ring:add_node(Ring, <<"slotB">>) == ok),

    ?assert(hash_ring:find_node(Ring, <<"keyA">>) == {ok, <<"slotA">>}),
    ?assert(hash_ring:find_node(Ring, <<"keyBBBB">>) == {ok, <<"slotA">>}),
    ?assert(hash_ring:find_node(Ring, <<"keyB_">>) == {ok, <<"slotB">>}),

    ?assert(hash_ring:remove_node(Ring, <<"slotA">>) == ok),
    ?assert(hash_ring:find_node(Ring, <<"keyA">>) == {ok, <<"slotB">>}),
    ?assert(hash_ring:find_node(Ring, <<"keyBBBB">>) == {ok, <<"slotB">>}).

## Benchmarks

The following benchmark was done on a MacBook Pro with the following specs:

* 2.8 Ghz Intel Core 2 Duo
* 8GB 1067 MHz DDR3
* Mac OS X 10.6.6

        ----------------------------------------------------
        bench: replicas = 1, nodes = 1, keys: 1000, ring size: 1
        ----------------------------------------------------
        generating and sorting 1 nodes...done
        generating keys...done
        stats: total = 0.04403s, avg/lookup: 0.44027us, min: 0.43374us, max: 0.73869us, ops/sec: 2272727
        ----------------------------------------------------
        bench: replicas = 1, nodes = 8, keys: 1000, ring size: 8
        ----------------------------------------------------
        generating and sorting 8 nodes...done
        generating keys...done
        stats: total = 0.04563s, avg/lookup: 0.45630us, min: 0.45339us, max: 0.50809us, ops/sec: 2192982
        ----------------------------------------------------
        bench: replicas = 1, nodes = 256, keys: 1000, ring size: 256
        ----------------------------------------------------
        generating and sorting 256 nodes...done
        generating keys...done
        stats: total = 0.04917s, avg/lookup: 0.49172us, min: 0.48684us, max: 0.54416us, ops/sec: 2036660
        ----------------------------------------------------
        bench: replicas = 8, nodes = 1, keys: 1000, ring size: 8
        ----------------------------------------------------
        generating and sorting 1 nodes...done
        generating keys...done
        stats: total = 0.04559s, avg/lookup: 0.45587us, min: 0.45276us, max: 0.46053us, ops/sec: 2197802
        ----------------------------------------------------
        bench: replicas = 8, nodes = 32, keys: 1000, ring size: 256
        ----------------------------------------------------
        generating and sorting 32 nodes...done
        generating keys...done
        stats: total = 0.04896s, avg/lookup: 0.48959us, min: 0.48601us, max: 0.54160us, ops/sec: 2044990
        ----------------------------------------------------
        bench: replicas = 8, nodes = 512, keys: 1000, ring size: 4096
        ----------------------------------------------------
        generating and sorting 512 nodes...done
        generating keys...done
        stats: total = 0.05401s, avg/lookup: 0.54006us, min: 0.53625us, max: 0.60199us, ops/sec: 1851852
        ----------------------------------------------------
        bench: replicas = 512, nodes = 8, keys: 1000, ring size: 4096
        ----------------------------------------------------
        generating and sorting 8 nodes...done
        generating keys...done
        stats: total = 0.05420s, avg/lookup: 0.54198us, min: 0.53301us, max: 1.03515us, ops/sec: 1848429
        ----------------------------------------------------
        bench: replicas = 512, nodes = 16, keys: 1000, ring size: 8192
        ----------------------------------------------------
        generating and sorting 16 nodes...done
        generating keys...done
        stats: total = 0.05576s, avg/lookup: 0.55763us, min: 0.55393us, max: 0.59292us, ops/sec: 1795332
        ----------------------------------------------------
        bench: replicas = 512, nodes = 32, keys: 100000, ring size: 16384
        ----------------------------------------------------
        generating and sorting 32 nodes...done
        generating keys...done
        stats: total = 5.73922s, avg/lookup: 0.57392us, min: 0.57195us, max: 0.58613us, ops/sec: 1745201
        ----------------------------------------------------
        bench: replicas = 512, nodes = 128, keys: 10000, ring size: 65536
        ----------------------------------------------------
        generating and sorting 128 nodes...done
        generating keys...done
        stats: total = 0.61953s, avg/lookup: 0.61953us, min: 0.61636us, max: 0.65255us, ops/sec: 1615509
        ----------------------------------------------------
        bench: replicas = 16, nodes = 1024, keys: 1000, ring size: 16384
        ----------------------------------------------------
        generating and sorting 1024 nodes...done
        generating keys...done
        stats: total = 0.05769s, avg/lookup: 0.57693us, min: 0.57307us, max: 0.63851us, ops/sec: 1736111



debian/changelog
hash-ring (1.0.0) stable; urgency=low

  * Initial release

 -- Chris Moos <chris@tech9computers.com>  Mon, 05 Mar 2012 02:37:00 -0700


hash_ring.c
/**
 * Copyright 2011 LiveProfile
 * 
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and 
 * limitations under the License.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <inttypes.h>

#include "sha1.h"
#include "hash_ring.h"
#include "sort.h"
#include "md5.h"

static int item_sort(void *a, void *b);

hash_ring_t *hash_ring_create(uint32_t numReplicas, HASH_FUNCTION hash_fn) {
    hash_ring_t *ring = NULL;
    
    // numReplicas must be greater than or equal to 1
    if(numReplicas <= 0) return NULL;
    
    // Make sure that the HASH_FUNCTION is supported
    if(hash_fn != HASH_FUNCTION_MD5 && hash_fn != HASH_FUNCTION_SHA1) return NULL;
    
    ring = (hash_ring_t*)malloc(sizeof(hash_ring_t));
    
    ring->numReplicas = numReplicas;
    ring->nodes = NULL;
    ring->items = NULL;
    ring->numNodes = 0;
    ring->numItems = 0;
    ring->hash_fn = hash_fn;
    ring->mode = HASH_RING_MODE_NORMAL;
    
    return ring;
}

void hash_ring_free(hash_ring_t *ring) {
    if(ring == NULL) return;

    // Clean up the nodes
    ll_t *tmp, *cur = ring->nodes;
    while(cur != NULL) {
        free(((hash_ring_node_t*)cur->data)->name);
        free(cur->data);
        tmp = cur;
        cur = tmp->next;
        free(tmp);
    }
    ring->nodes = NULL;
    
    // Clean up the items
    int x;
    for(x = 0; x < ring->numItems; x++) {
        free(ring->items[x]);
    }
    if(ring->items != NULL) free(ring->items);
    
    free(ring);
}

static int hash_ring_hash(hash_ring_t *ring, uint8_t *data, uint8_t dataLen, uint64_t *hash) {
    if(ring->hash_fn == HASH_FUNCTION_MD5) {
        uint8_t digest[16];
        md5_state_t state;
        md5_init(&state);
        
        md5_append(&state, (md5_byte_t*)data, dataLen);
        md5_finish(&state, (md5_byte_t*)&digest);

        if(ring->mode == HASH_RING_MODE_LIBMEMCACHED_COMPAT) {
            uint32_t low = (digest[3] << 24 | digest[2] << 16 | digest[1] << 8 | digest[0]);
            uint64_t keyInt;
            keyInt = low;
            *hash = keyInt;
            return 0;
        }
        else {
            uint32_t low = (digest[11] << 24 | digest[10] << 16 | digest[9] << 8 | digest[8]);
            uint32_t high = (digest[15] << 24 | digest[14] << 16 | digest[13] << 8 | digest[12]);
            uint64_t keyInt;
            
            keyInt = high;
            keyInt <<= 32;
            keyInt &= 0xffffffff00000000LLU;
            keyInt |= low;
            
            *hash = keyInt;
            
            return 0;
        }
    }
    else if(ring->hash_fn == HASH_FUNCTION_SHA1) {
        SHA1Context sha1_ctx;

        SHA1Reset(&sha1_ctx);
        SHA1Input(&sha1_ctx, data, dataLen);
        if(SHA1Result(&sha1_ctx) != 1) {
            return -1;
        }
        
        uint64_t keyInt = sha1_ctx.Message_Digest[3];
        keyInt <<= 32;
        keyInt |= sha1_ctx.Message_Digest[4];
        *hash = keyInt;
        return 0;
    }
    else {
        return -1;
    }
}

void hash_ring_print(hash_ring_t *ring) {
    if(ring == NULL) return;
    int x, y;
    
    printf("----------------------------------------\n");
    printf("hash_ring\n\n");
    printf("numReplicas:%8d\n", ring->numReplicas);
    printf("Nodes: \n\n");
    
    ll_t *cur = ring->nodes;
    x = 0;
    while(cur != NULL) {
        printf("%d: ", x);
        
        hash_ring_node_t *node = (hash_ring_node_t*)cur->data;
        
        for(y = 0; y < node->nameLen; y++) {
            printf("%c", node->name[y]);
        }
        printf("\n");
        cur = cur->next;
        x++;
    }
    printf("\n");
    printf("Items (%d): \n\n", ring->numItems);
    
    for(x = 0; x < ring->numItems; x++) {
        hash_ring_item_t *item = ring->items[x];
        printf("%" PRIu64 " : ", item->number);
        for(y = 0; y < item->node->nameLen; y++) {
            printf("%c", item->node->name[y]);
        }
        printf("\n");
    }
    
    printf("\n");
    printf("----------------------------------------\n");
    
}

int hash_ring_add_items(hash_ring_t *ring, hash_ring_node_t *node) {
    int x;
 
    char concat_buf[8];
    int concat_len;
    uint64_t keyInt;

    // Resize the items array
    void *resized = realloc(ring->items, (sizeof(hash_ring_item_t*) * ring->numNodes * ring->numReplicas));
    if(resized == NULL) {
        return HASH_RING_ERR;
    }
    ring->items = (hash_ring_item_t**)resized;
    for(x = 0; x < ring->numReplicas; x++) {
        if(ring->mode == HASH_RING_MODE_LIBMEMCACHED_COMPAT) {
            concat_len = snprintf(concat_buf, sizeof(concat_buf), "-%d", x);
        }
        else {
            concat_len = snprintf(concat_buf, sizeof(concat_buf), "%d", x);
        }

        uint8_t *data = (uint8_t*)malloc(concat_len + node->nameLen);
        memcpy(data, node->name, node->nameLen);
        memcpy(data + node->nameLen, &concat_buf, concat_len);

        if(hash_ring_hash(ring, data, concat_len + node->nameLen, &keyInt) == -1) {
            return HASH_RING_ERR;
        }
        
        hash_ring_item_t *item = (hash_ring_item_t*)malloc(sizeof(hash_ring_item_t));
        item->node = node;
        item->number = keyInt;
        
        ring->items[(ring->numNodes - 1) * ring->numReplicas + x] = item;
    }

    ring->numItems += ring->numReplicas;
    return HASH_RING_OK;
}

static int item_sort(void *a, void *b) {
    hash_ring_item_t *itemA = a, *itemB = b;
    if(a == NULL) return 1;
    if(b == NULL) return -1;

    if(itemA->number < itemB->number) {
        return -1;
    }
    else if(itemA->number > itemB->number) {
        return 1;
    }
    else {
        return 0;
    }
}

int hash_ring_add_node(hash_ring_t *ring, uint8_t *name, uint32_t nameLen) {
    if(ring == NULL) return HASH_RING_ERR;
    if(hash_ring_get_node(ring, name, nameLen) != NULL) return HASH_RING_ERR;
    if(name == NULL || nameLen <= 0) return HASH_RING_ERR;
    hash_ring_node_t *node = (hash_ring_node_t*)malloc(sizeof(hash_ring_node_t));
    if(node == NULL) {
        return HASH_RING_ERR;
    }
    
    node->name = (uint8_t*)malloc(nameLen);
    if(node->name == NULL) {
        free(node);
        return HASH_RING_ERR;
    }
    memcpy(node->name, name, nameLen);
    node->nameLen = nameLen;
    
    ll_t *cur = (ll_t*)malloc(sizeof(ll_t));
    if(cur == NULL) {
        free(node->name);
        free(node);
        return HASH_RING_ERR;
    }
    cur->data = node;

    // Add the node
    ll_t *tmp = ring->nodes;
    ring->nodes = cur;
    ring->nodes->next = tmp;
    
    ring->numNodes++;
    
    // Add the items for this node
    if(hash_ring_add_items(ring, node) != HASH_RING_OK) {
        hash_ring_remove_node(ring, node->name, node->nameLen);
        return HASH_RING_ERR;
    }

    // Sort the items
    quicksort((void**)ring->items, ring->numItems, item_sort);

    return HASH_RING_OK;
}

int hash_ring_remove_node(hash_ring_t *ring, uint8_t *name, uint32_t nameLen) {
    if(ring == NULL || name == NULL || nameLen <= 0) return HASH_RING_ERR;

    hash_ring_node_t *node;
    ll_t *next, *prev = NULL, *cur = ring->nodes;
    
    while(cur != NULL) {
        node = (hash_ring_node_t*)cur->data;
        if(node->nameLen == nameLen  &&
            memcmp(node->name, name, nameLen) == 0) {
                
                // Node found, remove it
                next = cur->next;
                free(node->name);
                
                if(prev == NULL) {
                    ring->nodes = next;
                }
                else {
                    prev->next = next;
                }
                
                int x;
                // Remove all items for this node and mark them as NULL
                for(x = 0; x < ring->numItems; x++) {
                    if(ring->items[x]->node == node) {
                        free(ring->items[x]);
                        ring->items[x] = NULL;
                    }
                }
                
                // By re-sorting, all the NULLs will be at the end of the array
                // Then the numItems is reset and that memory is no longer used
                quicksort((void**)ring->items, ring->numItems, item_sort);
                ring->numItems -= ring->numReplicas;
                
                free(node);
                free(cur);
                
                ring->numNodes--;
                
                return HASH_RING_OK;
        }
        
        prev = cur;
        cur = prev->next;
    }
    
    return HASH_RING_ERR;
}

hash_ring_node_t *hash_ring_get_node(hash_ring_t *ring, uint8_t *name, uint32_t nameLen) {
    if(ring == NULL || name == NULL || nameLen <= 0) return NULL;
    
    ll_t *cur = ring->nodes;
    while(cur != NULL) {
        hash_ring_node_t *node = (hash_ring_node_t*)cur->data;
        // Check if the nameLen is the same as well as the name
        if(node->nameLen == nameLen && 
            memcmp(node->name, name, nameLen) == 0) {
                return node;
        }
        cur = cur->next;
    }
    
    return NULL;
}

hash_ring_item_t *hash_ring_find_next_highest_item(hash_ring_t *ring, uint64_t num) {
    if(ring->numItems == 0) return NULL;
    
    int min = 0;
    int max = ring->numItems - 1;
    hash_ring_item_t *item = NULL;

    while(1) {
        if(min > max) {
            if(min == ring->numItems) {
                // Past the end of the ring, return the first item
                return ring->items[0];
            }
            else {
                // Return the next highest item
                return ring->items[min];
            }
        }
        
        int midpointIndex = (min + max) / 2;
        item = ring->items[midpointIndex];

        if(item->number > num) {
            // Key is in the lower half
            max = midpointIndex - 1;
        }
        else if(item->number <= num) {
            // Key is in the upper half
            min = midpointIndex + 1;
        }
    }
    
    return NULL;
}

hash_ring_node_t *hash_ring_find_node(hash_ring_t *ring, uint8_t *key, uint32_t keyLen) {
    if(ring == NULL || key == NULL || keyLen <= 0) return NULL;
    
    uint64_t keyInt;
    
    if(hash_ring_hash(ring, key, keyLen, &keyInt) == -1) return NULL;
    hash_ring_item_t *item = hash_ring_find_next_highest_item(ring, keyInt);
    if(item == NULL) {
        return NULL;
    }
    else {
        return item->node;
    }
}

int hash_ring_set_mode(hash_ring_t *ring, HASH_MODE mode) {
    if(ring == NULL) return HASH_RING_ERR;

    if(mode == HASH_RING_MODE_LIBMEMCACHED_COMPAT) {
        if(ring->hash_fn != HASH_FUNCTION_MD5) return HASH_RING_ERR;
        ring->mode = mode;
        return HASH_RING_OK;
    }
    else if(mode == HASH_RING_MODE_NORMAL) {
        ring->mode = mode;
        return HASH_RING_OK;
    }
    else {
        /* Invalid mode */
        return HASH_RING_ERR;
    }
}
